File: Level 3/베스트앨범/solution.js
function solution(genres, plays) {
  const answer = [];
  const genreSongs = new Map();
  const genrePlayTotalCount = new Map();

  for (let i = 0; i < genres.length; i++) {
    genreSongs.set(
      genres[i],
      genreSongs.get(genres[i])
        ? [...genreSongs.get(genres[i]), { id: i, play: plays[i] }]
        : [{ id: i, play: plays[i] }]
    );
    genrePlayTotalCount.set(
      genres[i],
      (genrePlayTotalCount.get(genres[i]) || 0) + plays[i]
    );
  }

  const sortedGenrePlayTotalCount = [...genrePlayTotalCount].sort(
    (a, b) => b[1] - a[1]
  );

  sortedGenrePlayTotalCount.forEach(([genre]) => {
    const twoSong = genreSongs
      .get(genre)
      .sort((a, b) => {
        if (a.play === b.play) {
          return a.id - b.id;
        }
        return b.play - a.play;
      })
      .map(({ id }) => id)
      .slice(0, 2);

    answer.push(...twoSong);
  });

  return answer;
}

File: Level 3/베스트앨범/solution.py
class Song:
    def __init__(self, id, play):
        self.id = id
        self.play = play


def solution(genres, plays):
    answer = []

    genre_dict = {}
    play_dict = {}

    for index, (genre, play) in enumerate(zip(genres, plays)):
        if genre not in genre_dict:
            genre_dict[genre] = []
            play_dict[genre] = 0

        genre_dict[genre].append(Song(index, play))
        play_dict[genre] += play

    sorted_genres = sorted(play_dict.items(), key=lambda x: x[1], reverse=True)

    for genre, _ in sorted_genres:
        sorted_songs = sorted(genre_dict[genre], key=lambda song: (-song.play, song.id))
        answer.extend(song.id for song in sorted_songs[:2])

    return answer
